quarkus-maven-plugin:create : Maven Wrapper installation fails with "SocketException: Connection reset", probably because of proxy

```
mvn io.quarkus:quarkus-maven-plugin:0.14.0:create \
    -DprojectGroupId=org.acme \
    -DprojectArtifactId=getting-started \
    -DclassName="org.acme.quickstart.GreetingResource" \
    -Dpath="/hello"
``` 
yields:
```
[INFO] --- quarkus-maven-plugin:0.14.0:create (default-cli) @ standalone-pom ---
Creating a new project in C:\Develop\_dev\quarkus\.
Configuration file created in src/main/resources/META-INF/application.properties
Downloading from releases: https://artifacts.somedomain.de/artifactory/prj-ccs-build-release/io/takari/maven-wrapper/0.5.3/maven-wrapper-0.5.3.pom
Downloaded from releases: https://artifacts.somedomain.de/artifactory/prj-ccs-build-release/io/takari/maven-wrapper/0.5.3/maven-wrapper-0.5.3.pom (2.4 kB at 8.8 kB/s)
Downloading from releases: https://artifacts.somedomain.de/artifactory/prj-ccs-build-release/io/takari/maven-wrapper/0.5.3/maven-wrapper-0.5.3.jar
Downloaded from releases: https://artifacts.somedomain.de/artifactory/prj-ccs-build-release/io/takari/maven-wrapper/0.5.3/maven-wrapper-0.5.3.jar (51 kB at 685 kB/s)
[ERROR] Unable to install the Maven wrapper (./mvnw) in the project
org.apache.maven.plugin.MojoExecutionException: Error installing the Maven Wrapper.
    at io.takari.maven.plugins.WrapperMojo.execute (WrapperMojo.java:94)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:119)
    at io.quarkus.maven.CreateProjectMojo.createMavenWrapper (CreateProjectMojo.java:186)
    at io.quarkus.maven.CreateProjectMojo.execute (CreateProjectMojo.java:163)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
Caused by: java.net.SocketException: Connection reset
    at java.net.SocketInputStream.read (SocketInputStream.java:210)
    at java.net.SocketInputStream.read (SocketInputStream.java:141)
    at sun.security.ssl.InputRecord.readFully (InputRecord.java:465)
    at sun.security.ssl.InputRecord.read (InputRecord.java:503)
    at sun.security.ssl.SSLSocketImpl.readRecord (SSLSocketImpl.java:975)
    at sun.security.ssl.SSLSocketImpl.performInitialHandshake (SSLSocketImpl.java:1367)
    at sun.security.ssl.SSLSocketImpl.startHandshake (SSLSocketImpl.java:1395)
    at sun.security.ssl.SSLSocketImpl.startHandshake (SSLSocketImpl.java:1379)
    at sun.net.www.protocol.https.HttpsClient.afterConnect (HttpsClient.java:559)
    at sun.net.www.protocol.https.AbstractDelegateHttpsURLConnection.connect (AbstractDelegateHttpsURLConnection.java:185)
    at sun.net.www.protocol.http.HttpURLConnection.getInputStream0 (HttpURLConnection.java:1564)
    at sun.net.www.protocol.http.HttpURLConnection.getInputStream (HttpURLConnection.java:1492)
    at sun.net.www.protocol.https.HttpsURLConnectionImpl.getInputStream (HttpsURLConnectionImpl.java:263)
    at org.apache.maven.wrapper.DefaultDownloader.downloadInternal (DefaultDownloader.java:90)
    at org.apache.maven.wrapper.DefaultDownloader.download (DefaultDownloader.java:76)
    at io.takari.maven.plugins.WrapperMojo.execute (WrapperMojo.java:78)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.twdata.maven.mojoexecutor.MojoExecutor.executeMojo (MojoExecutor.java:119)
    at io.quarkus.maven.CreateProjectMojo.createMavenWrapper (CreateProjectMojo.java:186)
    at io.quarkus.maven.CreateProjectMojo.execute (CreateProjectMojo.java:163)
    at org.apache.maven.plugin.DefaultBuildPluginManager.executeMojo (DefaultBuildPluginManager.java:137)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:210)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:156)
    at org.apache.maven.lifecycle.internal.MojoExecutor.execute (MojoExecutor.java:148)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:117)
    at org.apache.maven.lifecycle.internal.LifecycleModuleBuilder.buildProject (LifecycleModuleBuilder.java:81)
    at org.apache.maven.lifecycle.internal.builder.singlethreaded.SingleThreadedBuilder.build (SingleThreadedBuilder.java:56)
    at org.apache.maven.lifecycle.internal.LifecycleStarter.execute (LifecycleStarter.java:128)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:305)
    at org.apache.maven.DefaultMaven.doExecute (DefaultMaven.java:192)
    at org.apache.maven.DefaultMaven.execute (DefaultMaven.java:105)
    at org.apache.maven.cli.MavenCli.execute (MavenCli.java:956)
    at org.apache.maven.cli.MavenCli.doMain (MavenCli.java:288)
    at org.apache.maven.cli.MavenCli.main (MavenCli.java:192)
    at sun.reflect.NativeMethodAccessorImpl.invoke0 (Native Method)
    at sun.reflect.NativeMethodAccessorImpl.invoke (NativeMethodAccessorImpl.java:62)
    at sun.reflect.DelegatingMethodAccessorImpl.invoke (DelegatingMethodAccessorImpl.java:43)
    at java.lang.reflect.Method.invoke (Method.java:498)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launchEnhanced (Launcher.java:282)
    at org.codehaus.plexus.classworlds.launcher.Launcher.launch (Launcher.java:225)
    at org.codehaus.plexus.classworlds.launcher.Launcher.mainWithExitCode (Launcher.java:406)
    at org.codehaus.plexus.classworlds.launcher.Launcher.main (Launcher.java:347)
[INFO]
[INFO] ========================================================================================
[INFO] Your new application has been created in C:\Develop\_dev\quarkus\.
[INFO] Navigate into this directory and launch your application with mvn compile quarkus:dev
[INFO] Your application will be accessible on http://localhost:8080
[INFO] ========================================================================================
[INFO]
[INFO] ------------------------------------------------------------------------
[INFO] BUILD SUCCESS
[INFO] ------------------------------------------------------------------------
[INFO] Total time:  3.851 s
[INFO] Finished at: 2019-04-25T14:42:05+02:00
[INFO] ------------------------------------------------------------------------
```
This happens at work behind a proxy, but not at home (no proxy).

```
$ mvn -version
Apache Maven 3.6.1 (d66c9c0b3152b2e69ee9bac180bb8fcc8e6af555; 2019-04-04T21:00:29+02:00)
Maven home: ...
Java version: 1.8.0_202, vendor: Oracle Corporation, runtime: ...
Default locale: de_DE, platform encoding: Cp1252
OS name: "windows 10", version: "10.0", arch: "amd64", family: "windows"
```

Proxy environment variables are set/valid. Proxy is also defined in `settings.xml`.
I even tried adding `-Dhttp.proxyHost=...` to `MAVEN_OPTS`.

A quick look into the [proxy related issues of  takari/maven-wrapper](https://github.com/takari/maven-wrapper/issues?q=proxy) seems to suggest that quarkus needs to work around the maven-wrapper limitations or should at least document how to fix this manually.

--- a/devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java
+++ b/devtools/maven/src/main/java/io/quarkus/maven/CreateProjectMojo.java
@@ -36,6 +36,7 @@ import org.apache.maven.plugins.annotations.Parameter;
 import org.apache.maven.project.DefaultProjectBuildingRequest;
 import org.apache.maven.project.MavenProject;
 import org.apache.maven.project.ProjectBuilder;
+import org.apache.maven.settings.Proxy;
 import org.eclipse.aether.RepositorySystem;
 import org.eclipse.aether.RepositorySystemSession;
 import org.eclipse.aether.repository.RemoteRepository;
@@ -270,6 +271,8 @@ public class CreateProjectMojo extends AbstractMojo {
                     newExecutionRequest, session.getResult());
             newSession.setCurrentProject(newProject);
 
+            setProxySystemPropertiesFromSession();
+
             executeMojo(
                     plugin(
                             groupId("io.takari"),
@@ -285,6 +288,27 @@ public class CreateProjectMojo extends AbstractMojo {
         } catch (Exception e) {
             // no reason to fail if the wrapper could not be created
             getLog().error("Unable to install the Maven wrapper (./mvnw) in the project", e);
+        }
+    }
+
+    private void setProxySystemPropertiesFromSession() {
+        List<Proxy> proxiesFromSession = session.getRequest().getProxies();
+        // - takari maven uses https to download the maven wrapper
+        // - don't do anything if proxy system property is already set
+        if (!proxiesFromSession.isEmpty() && System.getProperty("https.proxyHost") == null) {
+
+            // use the first active proxy for setting the system properties
+            proxiesFromSession.stream()
+                    .filter(Proxy::isActive)
+                    .findFirst()
+                    .ifPresent(proxy -> {
+                        // note: a http proxy _is_ usable as https.proxyHost
+                        System.setProperty("https.proxyHost", proxy.getHost());
+                        System.setProperty("https.proxyPort", String.valueOf(proxy.getPort()));
+                        if (proxy.getNonProxyHosts() != null) {
+                            System.setProperty("http.nonProxyHosts", proxy.getNonProxyHosts());
+                        }
+                    });
         }
     }
 

